lib/matplotlib/pyplot.py

```diff
@@ -2069,11 +2069,17 @@ def set_cmap(cmap):
     matplotlib.cm.register_cmap
     matplotlib.cm.get_cmap
     """
-    cmap = cm.get_cmap(cmap)
+    cbook._check_isinstance((str, Colormap), cmap=cmap)
+    if isinstance(cmap, str):
+        name = cmap
+        cmap = get_cmap(name)
+        rc('image', cmap=name)
+    elif isinstance(cmap, Colormap):
+        if cmap.name not in cm.cmap_d.keys():
+            register_cmap(cmap.name, cmap)
+        rc('image', cmap=cmap.name)
```

Comment:
Here you could get into another tricky situation.

Say someone has *already* registered a different Colormap under the same name. So, the cmap.name is in the dictionary already, but it is actually pointing to a different colormap and this wouldn't set the global cmap to what you're expecting either.

Reply:
Ah, this is tricky. Perhaps:
```python
if cmap.name not in cm.cmap_d:
    register_cmap(cmap.name, cmap)
elif cmap is not get_cmap(cmap.name):
    raise some sort of error/warning
```
and in that error/warning explain that they should use `register_cmap` or change the name attribute of their colormap?

Also good to note that as originally written this function falls into the same trap. Though I suppose it's not super desirable to maintain backwards compatibility with a silent bug.

Reply:
Agreed, something like this would be desirable going forward to make it a little more clear what is going on and raising rather than silently doing something unexpected!